### ISSUE TITLE
Chapter 05 Distributed Git - some points

### ISSUE THREAD

**Issue by udyzr** [progit/progit2#1249]
**Which version of the book is affected?**  Version 2.1.146, 2019-04-15 pdf, source text, others not checked

**Describe the bug:**

Book is addressed to Git newbies. Also chapter 05 Distributed Git. The less book material student this sort has absolved the less points are obvious. In the beginning zero points are obvious. Thus it is critical to make precise word selection and accurate descriptions. Otherwise mental journey is put onto undesired trails. Way back to turn point might be costly.
There are few few points in chapter 5 Distributed Git not unambiguous enough. Lets review them. 
**Issue 1**
Contributing to a Project - Private Small Team, line number 159 - Description insinuates John had made two commits till that line's of text point of time. Please compare with elaborations shortly before first transcript for John's machine (line number 112).
So I propose to change to : Continuing with this example, shortly afterwards, John tries to push them to the same server:
Fixed form does not raise ambiguity.

**Steps to reproduce:**

**Expected behavior:**

**Screenshots:**

**Additional context:**

**Desktop:**

- Operating system:
- Browser/application:
- Browser/application version:

**Smartphone:**

- Device:
- OS:
- Browser/application:
- Browser/application version:

**E-book reader:**

- Device:
- Software Update:

**Comment by udyzr** [comment]
**Issue 2**
Contributing to a Project - Private Small Team, source text line number 273 reads: _Jessica can merge either origin/master or issue54 first — they’re both upstream..._
Open question: is for _master_ branch the branch _issue54_ really an upstream branch? In chapter about branching in Git I learned upstream branch is a branch being tracked by tracking branch. Here however _master_ and _issue54_ are two branches in the same repository.

**Comment by udyzr** [comment]
**Issue 3**
Contributing to a Project - Private Small Team, Figure 65 Simple multiple developer workflow
I miss for Jessica's timeline a loop accounting for her committing on _issu54_ branch. Commit series on issu54 branch then her intention to push them to upstream (on master) are the reasons why Jessica is merging.
So I propose to draw on Jessica's timeline two subsequent commit loops. The second loop following Jessica's push.

**Comment by ben** [comment]
General notes: 

* It would be helpful when reviewing these kinds of suggestions to include links to the relevant section on the git-scm.com website, or link to the line number you're talking about.
* If you're going to submit multiple issues, please submit them as separate issues, it helps keep the discussions focused.

## Issue 1

Referring to [this line](https://github.com/progit/progit2/blob/cb5a2d650d85f8eda93014d333f081054239539e/book/05-distributed-git/sections/contributing.asc#L159):

> So I propose to change to : Continuing with this example, shortly afterwards, John tries to push them to the same server:

This sounds fine to me. Want to submit a PR?

## Issue 2

The term "upstream" is sometimes ambiguous. In this context it means that Jessica wants to bring the changes from `origin/master` and `issue54` into her `master` branch. Maybe we could do a better job of defining this term?

## Issue 3

This figure isn't meant to be exhaustive. It's just meant to show the general sequence of fetches, merges, and pushes when developers are sharing a branch. I personally think it illustrates this nicely, and the reader fills in the elided steps themselves. Does anyone disagree? Should we add more detail to this diagram?

**Comment by GGGeuiwinegemPriceless** [comment]
I all the text are tainted I believe the json bug is prescent

**Comment by udyzr** [comment]
Sorry for delay, no free resources for matters here recently. Reg. Issue 1 OK, however I am not familiarized with building book from source text, toolchain, processes and terminology used here. PR?

**Comment by ben** [comment]
"PR" == "Pull request". Feel free to fork the repo, edit the file, commit on a branch, and submit a pull request. I'd love it if you got credit in the blame view for fixing this.

**Comment by udyzr** [comment]
Issue 1: PR raised

**Comment by udyzr** [comment]
Reg. Issue 2: Possibly it is not a this book specific problem. However we like to fix just this book. Defining the term might be full of efforts as checks will need to cover whole book. My own but current preference is to keep area of handling as small as possible - to try fixing it locally in this chapter section.

**Comment by ben** [comment]
I'm with you (I think) – "upstream" is ambiguous, and that causes problems. Despite its use in the docs, it's not really a technical term, but instead means "a place from which things flow to here." Why is `origin/master` upstream from `issue54`? Because we think of `issue54` as "subordinate" to `master`, and it will eventually merge back in. Like a tributary of a river.

At least that's the way I think of it. I'll grant you this probably makes less sense if you're not a native English speaker, and I'd _loooove_ to have a better way to describe these things, but "upstream" is so widely used that it might be hard to break away from.

**Comment by udyzr** [comment]
In my humble understanding these were repositories / projects where word "upstream" came to use first. All was fine as long as no other artifact started to use the same identifier - zero ambiguity. It is no problem if two artifacts are called using same name if former ones exist in different domains each - for repository and branch it is not the case.
Though I see you are with me if it concerns the direction of master to issue54 relation. By some way we consider master as issue54's upstream in contrast to book text's flow which sees issue54 as master's upstream.
I will look for how to remove word 'upstream' from this text.

**Comment by udyzr** [comment]
https://github.com/progit/progit2/blob/master/book/05-distributed-git/sections/contributing.asc#private-small-team, proceeding presented in lines 243 to 299 seem to be bit inconsequent. Jessica is merging `issue54`, then `origin/master` to `master` but prior to that she is checking the diff between `origin/master` and `issue54` (two branches being merged to `master`), means a diff between two branches not being merged together in direct way and explicit way. Would it not be more consequent to check instead diff between branches merge-from and merge-to? She is merging two branches to third one but checks diff between two former one?
In other words: how will the knowledge about diff between two merge-source branches help in merging them to third one merge-destination branch?

**Comment by ben** [comment]
@schacon wrote this section, so I'd defer to him regarding the intent, but in a way it seems arbitrary. If you have a divergent `master` and a feature branch, and you want to push them both to `origin`, you have two choices:

1. `git checkout master && git merge origin/master && git merge feature`
1. `git checkout master && git merge feature && git merge origin/master`

Either way, you'll end up with the same tree in the final commit. Maybe Jessica is trying to figure out which way is easier? Maybe she just wants to review what's being merged? There are a number of reasons you'd want to do this, and the book is just illustrating one way of going about it.

**Comment by udyzr** [comment]
We are discussing here not the order of two merges in sequence being performed.
The point is which diff is being looked at prior performing those two merges.
I wonder what might be the value of checking diff between two branches before merging them to third branch. I also wonder for what reason Jessica apparently might not be interested in knowing diff to particular branch being merged to prior to performing these two merges.
> Maybe Jessica is trying to figure out which way is easier? Maybe she just wants to review what's being merged?
She is not checking the distances to master so I wonder how will her proceeding help her to see which of those two ways is easier. She is also not reviewing what's is being merged. She is reviewing set of commits located between two branches being merged to third branch.
Addressed piece of book's text raises above questions then leaves them open.

**Comment by udyzr** [comment]
Actually I am wrong with my judgment regarding this question.

**Comment by ben** [comment]
OK, going to close this out then.
